src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceValueSerializer.cs

```diff
@@ -0,0 +1,156 @@
+        }
+
+        /// <inheritdoc/>
+        public sealed override ODataValue CreateODataValue(object graph, IEdmTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            if (!expectedType.IsStructured())
+            {
+                throw Error.InvalidOperation(SRResources.CannotWriteType, typeof(ODataResourceValueSerializer), expectedType.FullName());
+            }
+
+            ODataResourceValue value = CreateODataResourceValue(graph, expectedType.AsStructured(), writeContext);
+            if (value == null)
+            {
+                return new ODataNullValue();
+            }
+
+            return value;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1804:RemoveUnusedLocals", MessageId = "edmTypeSerializer")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic")]
+        private ODataResourceValue CreateODataResourceValue(object graph, IEdmStructuredTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            List<ODataProperty> properties = new List<ODataProperty>();
+            ODataResourceValue resourceValue = new ODataResourceValue();
+
+            IDelta delta = graph as IDelta;
+            if (delta != null)
+            {
+                foreach (string propertyName in delta.GetChangedPropertyNames())
+                {
+                    object propertyValue;
+                    delta.TryGetPropertyValue(propertyName, out propertyValue);
+
+                    if (propertyValue != null)
+                    {
+                        SetPropertyValue(writeContext, properties, propertyName, propertyValue);
+                    }
+                }
+
+                foreach (string propertyName in delta.GetUnchangedPropertyNames())
+                {
+                    object propertyValue;
+                    delta.TryGetPropertyValue(propertyName, out propertyValue);
+
+                    if (propertyValue != null)
+                    {
+                        SetPropertyValue(writeContext, properties, propertyName, propertyValue);
+                    }
+                }
+            }
+            else
+            {
+                resourceValue.TypeName = expectedType.FullName();
+
+                foreach (PropertyInfo property in graph.GetType().GetProperties())
+                {
+                    object propertyValue = property.GetValue(graph);
+
+                    SetPropertyValue(writeContext, properties, property.Name, propertyValue);
+                }
+            }            
+
+            resourceValue.Properties = properties;
+
+            return resourceValue;
+        }
+
+        private void SetPropertyValue(ODataSerializerContext writeContext, List<ODataProperty> properties,string propertyName, object propertyValue)
+        {
+            if (propertyValue != null)
```

Comment:
it should be valid to serialize a null value. The serializers for nullable types should all support passing a propertyValue of null, and return ODataNullValue.